pkg/apis/monitoring/v1/thanos_types.go

```diff
@@ -0,0 +1,134 @@
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Specification of the desired behavior of the ThanosRuler cluster. More info:
+	// https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	Spec ThanosRulerSpec `json:"spec"`
+	// Most recent observed status of the ThanosRuler cluster. Read-only. Not
+	// included when requesting from the apiserver, only from the ThanosRuler
+	// Operator API itself. More info:
+	// https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	Status *ThanosRulerStatus `json:"status,omitempty"`
+}
+
+// ThanosRulerList is a list of ThanosRulers.
+// +k8s:openapi-gen=true
+type ThanosRulerList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata
+	// More info: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ListMeta `json:"metadata,omitempty"`
+	// List of Prometheuses
+	Items []*ThanosRuler `json:"items"`
+}
+
+// ThanosRulerSpec is a specification of the desired behavior of the ThanosRuler. More info:
+// https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+// +k8s:openapi-gen=true
+type ThanosRulerSpec struct {
+	// Standard object’s metadata. More info:
+	// https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// Metadata Labels and Annotations gets propagated to the prometheus pods.
+	PodMetadata *metav1.ObjectMeta `json:"podMetadata,omitempty"`
+	// Thanos container image URL.
+	Image string `json:"image,omitempty"`
+	// An optional list of references to secrets in the same namespace
+	// to use for pulling thanos images from registries
+	// see http://kubernetes.io/docs/user-guide/images#specifying-imagepullsecrets-on-a-pod
+	ImagePullSecrets []v1.LocalObjectReference `json:"imagePullSecrets,omitempty"`
+	// When a ThanosRuler deployment is paused, no actions except for deletion
+	// will be performed on the underlying objects.
+	Paused bool `json:"paused,omitempty"`
+	// Number of thanos ruler instances to deploy.
+	Replicas *int32 `json:"replicas,omitempty"`
+	// Resources defines the resource requirements for the Thanos sidecar.
+	// If not provided, no requests/limits will be set
+	Resources v1.ResourceRequirements `json:"resources,omitempty"`
+	// ObjectStorageConfig configures object storage in Thanos.
+	ObjectStorageConfig *v1.SecretKeySelector `json:"objectStorageConfig,omitempty"`
+	// ListenLocal makes the Thanos sidecar listen on loopback, so that it
+	// does not bind against the Pod IP.
+	ListenLocal bool `json:"listenLocal,omitempty"`
+	// QueryEndpoints defines Thanos querier endpoints from which to query metrics.
+	// Maps to the --query flag of thanos ruler.
+	QueryEndpoints []string `json:"queryEndpoints"`
```

Comment:
This seems to be a required field right now.

I'm just guessing but maybe this makes it optional
```suggestion
	QueryEndpoints *[]string `json:"queryEndpoints"`
```

Reply:
It's required to have some value when starting the container.  This field just translates to `--query` arguments to the thanos binary, and at least one is required when running `thanos rule`.  I guess this is because without any query endpoint you could never evaluate any rules.

Reply:
Correct though it has changed a bit with the addition of the `--query-config*` flags which supersede `--query`.

https://github.com/thanos-io/thanos/blob/46a97fdc9765bbe7eb9d75d801abf186015e1b01/cmd/thanos/rule.go#L148-L150